Receiver: Use proxy store for MultiTSDB

matej-g opened the pull request:
<!--
    Keep PR title verbose enough and add prefix telling
    about what components it touches e.g "query:" or ".*:"
-->

<!--
    Don't forget about CHANGELOG!

    Changelog entry format:
    - [#<PR-id>](<PR-URL>) Thanos <Component> ...

    <PR-id> Id of your pull request.
    <PR-URL> URL of your PR such as https://github.com/thanos-io/thanos/pull/<PR-id>
    <Component> Component affected by your changes such as Query, Store, Receive.
-->

* [ ] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

This PR attempts to address a long-standing issue https://github.com/thanos-io/thanos/issues/2864.

The PR proposes a couple of changes for the solution to work:
- It adjusts the `Client` interface to replace `Addr` method with method `StoreInfo`. The aim of this is to accommodate for both cases of clients - remote clients (i.e. the typical store client that proxy is dialing remotely) and local clients (this are clients where proxy won't be making an external call - so in our case the `MultiTSDB` - but rather we'll use obtain the series 'locally' from individual tenant TSDBs). `StoreInfo` shall return information about store client, namely: a) is this a remote or local client; b) if this client is remote, return it's address as well. Just as previously, these information can be then used e.g. for logging and debugging store connections.
- Add `localClient` implementation of `proxy.Client` to MultiTSDB.
- Use `ServerAsClient` implementation of the `StoreClient` gRPC interface, to obtain series. This shall replace the existing snowflake solution with an implementation of `proxy.Client`.

## Verification

Test have been adjust and are running succesfully.

I have also done a few rounds with the interactive test and everything seems to be working as expected.

I'd welcome feedback on the idea of local / remote client (whether this is a good adjustment to the interface) as well as feedback on using server-as-client, whether there are any drawbacks or potential issues with using this store client implementation (I don't think it's used anywhere besides testing).

bwplotka commented on pkg/store/proxy.go:
```diff
@@ -37,6 +37,16 @@ type ctxKey int
+// StoreType determines if the store for a given client is a local store
+// (i.e. using server-as-client with in-process stream) or a store to which
+// the client connects remotely over gRPC.
+type StoreType string
```
Why we need to differntiate?

matej-g replied:
In line with discussion in https://github.com/thanos-io/thanos/issues/2864, I also felt the current interface does fit that good with the local client scenario. Currently, when calling `Addr()` we should get a valid `host:port` address. 

In case of local client, however, there is realistically no address - we're not calling any address, we're not calling `localhost`, we're using the server-as-client implementation, meaning we're not really calling any address but directly streaming the series. 

So we could hide this away and make up an address, but that does not seem right to me since there is not really an address attached to the local client. Maybe other alternative would be to not return any address or return some other string (e.g. `<local_client>`) to indicate that the address is not applicable. Then we could avoid having this type but it seems still somewhat less clear than having this explicitly stated.

I'm open to this or other alternative solutions / opinions though.

bwplotka replied:
Yea, let me check next week, but I see your point. Another way would be to create `Addr() (addr string, noAddress bool)` which is simple and self-explanatory.

matej-g replied:
That could also work :+1: I'm open to whatever is best understandable for others :blush:

matej-g replied:
@fpetkovski pretty much, filtering and I guess debugging, tracing tags etc.

I see your points, after more thinking I'd agree that it might be overkill to have that type since it's basically either one or the other, so something along what @bwplotka suggested should be clear enough, I'll remake this.

bwplotka replied:
Sounds like we need to address it still.

matej-g replied:
Addressed :+1:

bwplotka commented on pkg/store/proxy.go:
```diff
@@ -591,12 +604,17 @@ func storeMatchDebugMetadata(s Client, storeDebugMatchers [][]*labels.Matcher) (
 
+	clientType, addr := s.StoreInfo()
+	if clientType != Remote {
+		return false, "the store is not remote, cannot match __address__"
```
Hmm, you can - just it can be `local` or some derivative like `local-tenant1`, no? (:

matej-g replied:
Similar to my comment above, I understand `__address__` to be applicable really only to remote stores.

bwplotka reviewed (commented):
Looks good generally, thanks! 

Just a question about local/remote - does it needs to separate method to distinguish? Why not just hide this fact in the implementation of the client and return the `localhost-<tenantX>` address? 🤔 

Perhaps the implementation might be simpler if we remove this complexity and teat "local" stores (someday store/proxy  might be just endpoint proxy) as simply store at the `localhost-tenant`. Essentially sentinel addresses 🤔

fpetkovski commented:
@matej-g would you mind rebasing this PR? I'd like to do another review.

matej-g commented:
> @matej-g would you mind rebasing this PR? I'd like to do another review.

Phew, done, thanks for taking a look! :slightly_smiling_face:

fpetkovski reviewed (commented):
This would be awesome to get merged in light of https://github.com/thanos-io/thanos/pull/5692#discussion_r972855049

matej-g commented on cmd/thanos/receive.go:
```diff
@@ -306,11 +306,14 @@ func runReceive(
 			comp,
-			dbs.TSDBStores,
+			labels.Labels{},
+			10*time.Second,
```
Thanks for catching this, I actually wanted to change this and disable timeout (make it `0`) since there's no network call involved, so I believe it should be fine.

fpetkovski replied:
So we're not changing this?

matej-g replied:
I'm about to push, sorry for the delays

matej-g commented on pkg/receive/multitsdb.go:
```diff
@@ -88,6 +90,45 @@ func NewMultiTSDB(
+}
+
+func (l *localClient) SupportsSharding() bool {
+	return false
```
Right, this should be the opposite, thanks. Do we actually have stores that do not support sharding?

fpetkovski replied:
We don't have production stores without sharding support. It's mostly here for backwards compatibility, for test stubs, and for external implementations of the store api :)

bwplotka replied:
Can you remind me what this changes?

fpetkovski replied:
Sharding will be done in the querier if this flag is set to false. It will be a no-op, but still we will have to hashmod all series.

matej-g commented on cmd/thanos/receive.go:
```diff
@@ -306,11 +306,14 @@ func runReceive(
-			dbs.TSDBStores,
+			labels.Labels{},
+			10*time.Second,
+			store.EagerRetrieval,
```
I haven't watched the recent changes too closely, but my understanding was that lazy retrieval strategy is still experimental feature, I went therefore with what is also default in querier now.

bwplotka replied:
Both fine, but probably lazy better as we aim for lazy for future and there are no blockers for it. 

On querier lazy is not needed with the current PromQL engine which iterates over all series no matter what.

matej-g replied:
Changing to lazy then as there are no blockers.

matej-g replied:
Hm so changing to lazy seems to have some issues, causes receiver to panic, I'll take a look. For now I'll leave it at eager.

matej-g replied:
This has been changed after the nil fix, all looks good, resloving.

bwplotka commented on pkg/query/endpointset.go:
```diff
@@ -771,11 +766,14 @@ func (er *endpointRef) SupportsSharding() bool {
+	)
 }
 
-func (er *endpointRef) Addr() string {
```
Right, but as @matej-g wrote Addr method does not make sense for an abstraction that works on both local and remote endpoint, so I feel his proposal makes sense (plus is relevant to the PR) (:

fpetkovski replied:
Right, but that method does not need to be part of the interface :) Anyway, let's keep it as it is since the PR has been in review for a while now.

bwplotka reviewed (approved):
Looks generally good, just a question about timeouts and that StoreType. It also overlaps with "component type" a lot we use for Store API.

Thanks!

matej-g commented:
Hey @fpetkovski @bwplotka,

Thanks for the reviews, I addressed all of the comments, except for using lazy retrieval strategy in proxy, since it seem to be causing receiver to panic. If we're happy to move forward with eager strategy for now, I can take a look and figure it our on the side.

fpetkovski reviewed (approved):
Thanks for this awesome work!

matej-g commented:
I'll wait on https://github.com/thanos-io/thanos/pull/5717 and switch the eval strategy, I believe then this is finally good to go :+1:

GiedriusS commented:
Could you please double check if https://github.com/thanos-io/thanos/pull/5746 fixes the `nil` panic?

yeya24 commented:
This seems like to be closed accidentally by Github, reopen now.

matej-g commented:
Sorry for the delay, PR is now updated, this should be good to go once CI passes

fpetkovski reviewed (approved):
LGTM on green!

matej-g commented:
The check for `Go build for different platform` on the PR is not getting out of `Expected` state, but the job actually ran succesfully (https://github.com/thanos-io/thanos/actions/runs/3195252383/jobs/5215692706).

I'm not sure if this is due to yesterday's webhook issues on GH side, but given this is approved and the checks actually ran successfully, I'm merging.
